kotlin-extensions/src/main/java/com/amplifyframework/ktx/DataStoreCategoryExt.kt

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+@file:Suppress("unused")
+
+package com.amplifyframework.ktx // TODO: or .datastore?
+
+import com.amplifyframework.core.async.Cancelable
+import com.amplifyframework.core.model.Model
+import com.amplifyframework.core.model.query.QueryOptions
+import com.amplifyframework.core.model.query.predicate.QueryPredicate
+import com.amplifyframework.datastore.DataStoreCategory
+import com.amplifyframework.datastore.DataStoreException
+import com.amplifyframework.datastore.DataStoreItemChange
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.sendBlocking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+@Throws(DataStoreException::class)
+suspend fun <T: Model> DataStoreCategory.save(item: T) {
+    return suspendCoroutine { continuation ->
+        save(item,
+            { continuation.resume(Unit) },
+            { continuation.resumeWithException(it) }
+        )
+    }
+}
+
+@Throws(DataStoreException::class)
+suspend fun <T: Model> DataStoreCategory.save(item: T, predicate: QueryPredicate) {
+    return suspendCoroutine { continuation ->
+        save(item, predicate,
+            { continuation.resume(Unit) },
+            { continuation.resumeWithException(it) }
+        )
+    }
+}
+
+@Throws(DataStoreException::class)
+suspend fun <T: Model> DataStoreCategory.delete(item: T) {
+    return suspendCoroutine { continuation ->
+        delete(item,
+            { continuation.resume(Unit) },
+            { continuation.resumeWithException(it) }
+        )
+    }
+}
+
+@Throws(DataStoreException::class)
+suspend fun <T: Model> DataStoreCategory.delete(item: T, predicate: QueryPredicate) {
+    return suspendCoroutine { continuation ->
+        delete(item, predicate,
+            { continuation.resume(Unit) },
+            { continuation.resumeWithException(it) }
+        )
+    }
+}
+
+@Throws(DataStoreException::class)
+suspend fun <T: Model> DataStoreCategory.query(clazz: Class<T>): Iterator<T> {
```

Comment:
You can also do something like:

```kt
suspend inline fun <reified T: Model> DataStoreCategory.query(): Iterator<T> {
    return suspendCoroutine { continuation ->
        query(
            T::class.java,
            { continuation.resume(it) },
            { continuation.resumeWithException(it) }
        )
    }
}
```

Which allows callers to do:
```
val results: Iterator<MyItem> = store.query()
// OR
val results = store.query<MyItem>()
```


See: https://kotlinlang.org/docs/reference/inline-functions.html#reified-type-parameters

Reply:
In the current version of the PR, I have updated `query()` to return a `Flow<T>`.

I do still have the `Class<T>` as an argument. My concern is that `storage.query()` isn't explicit enough about the user's intention, since the behavior can change depending on the _implicit_ type.

This would be a good topic to discuss in design review.